### PR TITLE
Make ExchangeSampler only propose different configurations

### PR DESCRIPTION
**Pull request by PhilipVinc**
This was implemented by @jwnys for Fermionic states, but makes sense in general for the ExchangeRule in general.

### CODE REVIEW

File: netket/sampler/rules/exchange.py [hunk 1]
```diff
@@ -139,3 +157,12 @@ def compute_clusters(graph: AbstractGraph, d_max: int):
         res_clusters[i] = np.asarray(cluster)
 
     return res_clusters
+
+
+@jax.jit
+def _compute_different_clusters_mask(clusters, σ):
```
Comment (jwnys) [inline netket/sampler/rules/exchange.py:163]: Here we could maybe check 
```python
if jnp.issubdtype(σ.dtype, int):
    hoppable_clusters_mask = σ[..., clusters[:, 0]] != σ[..., clusters[:, 1]]
else:
    hoppable_clusters_mask = ~jnp.isclose(
        σ[..., clusters[:, 0]], σ[..., clusters[:, 1]]
    )

**Review by jwnys — CHANGES_REQUESTED** [review]
Looks good

**Comment by jwnys** [comment]
Thinking a bit more about this, maybe we should impose some more checks?
- For bosons or Fock space different than the fermionic one: if this is a single boson move or "exchange", an exchange would check something different.
- For higher spin components, there are more moves allowed that maintain the Sz that should be considered.

**Review by gcarleo — CHANGES_REQUESTED** [review]
if I understand correctly this does random permutations of the local occupations ? however this is not general,we should rather do a random choice of the local two site configurations such that the total number is locally conserved.(unless I misunderstood the implementation)

**Comment by gcarleo** [comment]
yes but these simple exchange moves are not ergodic, for local Hilbert spaces with more than two states, not sure we should have a warning in those cases ?

**Comment by PhilipVinc** [comment]
That's already the case...

**Comment by PhilipVinc** [comment]
As for warnings, yes, I've long agreed.
As usual someone must do it

**Comment by gcarleo** [comment]
ok, but yes I actually agree that being an exchange sampler the
implementation you are using is ok. we should add warnings more generally
but indeed it's a lot more work because we need to touch all samplers and
check consistency. The more general MetropolisHop or similar has already
been implemented by Zakari, maybe he could do a PR with that

**Review by gcarleo — APPROVED** [review]

**Comment by PhilipVinc** [comment]
I think that having a proper 'hopping' rule is the way to go.
With possibility to 'hop' 1,2 or N values at a time.

We can keep this exchange kernel around for local size 2 (because it is simpler and probably faster) and as an example, but also because it might make sense to use it.

**Pull request merged**